src/app/shared/copier/copier.service.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * This class is based on the code in the following projects:
+ *
+ * - https://github.com/zenorocha/select
+ * - https://github.com/zenorocha/clipboard.js/
+ *
+ * Both released under MIT license - © Zeno Rocha
+ */
+import {Injectable} from '@angular/core';
+
+@Injectable()
+export class CopierService {
+  private fakeElem: HTMLTextAreaElement;
+
+  /**
+   * Creates a fake textarea element, sets its value from `text` property,
+   * and makes a selection on it.
+   */
+  createFake(text: string) {
+    const isRTL = document.documentElement.getAttribute('dir') === 'rtl';
+
+    // Create a fake element to hold the contents to copy
+    this.fakeElem = document.createElement('textarea');
+
+    // Prevent zooming on iOS
+    this.fakeElem.style.fontSize = '12pt';
+
+    // Reset box model
+    this.fakeElem.style.border = '0';
+    this.fakeElem.style.padding = '0';
+    this.fakeElem.style.margin = '0';
+
+    // Move element out of screen horizontally
+    this.fakeElem.style.position = 'absolute';
+    this.fakeElem.style[ isRTL ? 'right' : 'left' ] = '-9999px';
+
+    // Move element to the same position vertically
+    const yPosition = window.pageYOffset || document.documentElement.scrollTop;
+    this.fakeElem.style.top = yPosition + 'px';
+
+    this.fakeElem.setAttribute('readonly', '');
+    this.fakeElem.value = text;
+
+    document.body.appendChild(this.fakeElem);
+
+    this.fakeElem.select();
+    this.fakeElem.setSelectionRange(0, this.fakeElem.value.length);
+  }
+
+  removeFake() {
+    if (this.fakeElem) {
+      document.body.removeChild(this.fakeElem);
+      this.fakeElem = null;
+    }
+  }
+
+  copyText(text: string) {
+    try {
```

Comment:
Under what circumstances would this throw an error?

https://github.com/angular/material2/blob/master/CODING_STANDARDS.md#try-catch

Reply:
This service is straight from [core](https://github.com/angular/angular/blob/master/aio/src/app/shared/copier.service.ts). It looks like the try/catch originally comes from [here](https://github.com/zenorocha/clipboard.js/commit/231d62db0e64e8eaeea3bf3b6627418f1ced23ab), but as far as I can tell, that was written before `execCommand` was supported by all major browsers. Will remove.

Reply:
I have higher code quality standards than core :wink: